program/src/processor/start.rs

```diff
@@ -1,18 +1,107 @@
+use mpl_token_metadata::state::CollectionAuthorityRecord;
+
 use super::*;
 
-pub fn start_migration(_program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
-    return Err(MigrationError::FeatureDisabled.into());
+pub fn start_migration(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
     // Fetch accounts
     let account_info_iter = &mut accounts.iter();
+    let payer_info = next_account_info(account_info_iter)?;
     let authority_info = next_account_info(account_info_iter)?;
     let collection_mint_info = next_account_info(account_info_iter)?;
     let collection_metadata_info = next_account_info(account_info_iter)?;
+    let delegate_info = next_account_info(account_info_iter)?;
+    let delegate_record_info = next_account_info(account_info_iter)?;
     let migration_state_info = next_account_info(account_info_iter)?;
+    let spl_token_program_info = next_account_info(account_info_iter)?;
     let system_program_info = next_account_info(account_info_iter)?;
+    let _token_metadata_program_info = next_account_info(account_info_iter)?;
 
-    // Validate Accounts
-
+    // Authority is a signer
     assert_signer(authority_info)?;
 
+    // The migration state account must must match the correct derivation
```

Comment:
Yeah, the current handler is not susceptible to this attack because the authority has to be a signer, and the authority is checked against the authority on the migration state. If someone has the private key for the pubkey of all zeros we have a bigger problem on our hands. 

Adding an owner check here doesn't help either, because a migration state account with all zeros can still be assigned to the `mpl-migration-validator` program. However, to prevent issues with regression in the future and as a nice general addition to our `from_account_info` method I've added a check for zeroed data:

```rust

if data.is_zeroed() {
    return Err(DeserializationError::EmptyMigrationState.into());
}
```

using a custom trait implemented for a slice of u8:

```rust
trait Zeroed {
    fn is_zeroed(&self) -> bool;
}


impl Zeroed for [u8] {
    fn is_zeroed(&self) -> bool {
        self.iter().all(|&x| x == 0)
    }
}
```

This might be a nice addition to our `TokenMetadataAccount` trait as another layer of safety in account checking.

Reply:
I also added a test to trigger the new error.

Reply:
Cool.  It might be better to steal the implementation from solana that does it in chunks.  I'm not sure about it on a blockchain, but if I just compare these two for time:
```
fn is_zeroed_byte_by_byte(data: &[u8]) -> bool {
    data.iter().all(|&x| x == 0)
}

/// See if a slice contains all zeroes.  Useful for checking an account's data.
pub fn is_zeroed_chunks(buf: &[u8]) -> bool {
    const ZEROS_LEN: usize = 1024;
    const ZEROS: [u8; ZEROS_LEN] = [0; ZEROS_LEN];
    let mut chunks = buf.chunks_exact(ZEROS_LEN);

    #[allow(clippy::indexing_slicing)]
    {
        chunks.all(|chunk| chunk == &ZEROS[..])
            && chunks.remainder() == &ZEROS[..chunks.remainder().len()]
    }
}
```

I get:
```
3000 byte by byte: 102.333µs
3000 in chunks: 1.041µs
```